src/dispatch/batch_builder.rs

```diff
@@ -0,0 +1,296 @@
+
+        let stage = self.stages_builder.build();
+
+        // Average
+        let batch_running_time = self.running_time / (self.system_count as f32);
+        // Depending on the system count increase the running time
+        let batch_running_time = batch_running_time + 2.0 * (self.system_count as f32 / 15.0);
+        let batch_running_time = batch_running_time.round() as i32;
+
+        let batch_running_time = match batch_running_time {
+            1 => RunningTime::VeryShort,
+            2 => RunningTime::Short,
+            3 => RunningTime::Average,
+            4 => RunningTime::Long,
+            _ => RunningTime::VeryLong,
+        };
+
+        #[cfg(feature = "parallel")]
+        let res = (self.read_resources, self.write_resources, Batch::new(stage, batch_running_time, self.thread_pool.unwrap()));
+
+        #[cfg(not(feature = "parallel"))]
+        let res = (self.read_resources, self.write_resources, Batch::new(stage, batch_running_time));
+
+        res
+    }
+}
+
+
+pub struct Batch<'a> {
+    stages: Vec<Stage<'a>>,
+    running_time: RunningTime,
+
+    #[cfg(feature = "parallel")]
+    thread_pool: ::std::sync::Arc<::rayon::ThreadPool>,
+}
+
+impl<'a> Batch<'a> {
+    #[cfg(feature = "parallel")]
+    pub(crate) fn new(stages: Vec<Stage<'a>>, running_time: RunningTime, thread_pool: ::std::sync::Arc<::rayon::ThreadPool>) -> Self {
+        Batch{
+            stages,
+            running_time,
+            thread_pool,
+        }
+    }
+
+    #[cfg(not(feature = "parallel"))]
+    pub(crate) fn new(stages: Vec<Stage<'a>>, running_time: RunningTime) -> Self {
+        Batch{
+            stages,
+            running_time,
+        }
+    }
+
+    pub fn running_time(&self) -> RunningTime {
+        self.running_time
+    }
+
+    /// Dispatches the systems (except thread local systems)
+    /// in parallel given the resources to operate on.
+    ///
+    /// This operation blocks the
+    /// executing thread.
+    ///
+    /// Only available with "parallel" feature enabled.
+    ///
+    /// Please note that this method assumes that no resource
+    /// is currently borrowed. If that's the case, it panics.
+    #[cfg(feature = "parallel")]
+    pub fn dispatch_par(&mut self, world: &World) {
+        let stages = &mut self.stages;
+
+        self.thread_pool.install(move || {
+            for stage in stages {
+                stage.execute(world);
+            }
+        });
+    }
+
+    /// Dispatches the systems (except thread local systems) sequentially.
+    ///
+    /// This is useful if parallel overhead is
+    /// too big or the platform does not support multithreading.
+    ///
+    /// Please note that this method assumes that no resource
+    /// is currently borrowed. If that's the case, it panics.
+    pub fn dispatch_seq(&mut self, world: &World) {
+        for stage in &mut self.stages {
+            stage.execute_seq(world);
+        }
+    }
+}
+
+impl<'a> RunNow<'a> for Batch<'a>{
```

Comment:
`RunNow` is not sufficient if you want to add `Batch` to a dispatcher.

Reply:
I'm going to work on this soon today.

The solution to this problem is not use the batch system and register the systems under the BatchBuilder directly inside the stages vector.

The problem with this is that using directly the BatchSystem is much more convenient since their systems can't be mixed with the other systems. Also its restart happen locally and no further changes are required.

Instead the solution of placing the batching systems directly inside the stages vector make necessary keep track of them in order to not mix these with the other systems.
This is necessary because a normal system may be executed many times unexpectedly.

What's necessary to change in order to allow to add an object which implement just the RunNow  trait?